[patch] Refactor Grammar

mmaloney-sf opened the pull request:
As far as I can tell, the grammar for FIRRTL was originally created with two design points:

- Make a pretty, whitespace-sensitive language.
- Just make it work.

The current grammar is overgrown and in need of a refactoring. I have made several changes to help illuminate the structure of the FIRRTL language:

- The organization of the production rules has been tightened up a bit
- Several grammar rules have been renamed for uniformity or clarity.
-The `expr` production rule  has been change to an alternate list of non-terminals
-The `statement` production rule  has been change to an alternate list of non-terminals

The `expr` and `statement` production rules now allow you to glance at the grammar and get a reasonable idea of what a language-level AST should look like. For `statement` especially, you can see the taxonomy of statements more clearly: there are circuit components, connect-like statements, conditional blocks, commands, groups, and the `skip` statement.

While I was careful to keep this grammar equivalent, it's hard for me to verify. Please review these changes carefully, and ask to discuss any changes that seem questionable. The PR is best reviewed commit-by-commit. Even if the grammar has minor mistakes, the increase in clarity should be a net-positive.

I have another branch where I make two additional, more controversial changes:
- the removal of the lexical rules for tokens (ie, `dedent` is already not defined, so why not also leave `int` `string` `id` and `info` undefined as well?)
- the replacement of `primop` with a much simpler variant that doesn't bake the primitive operations into the grammar. (This requires elaboration on when a `primop` expression is well-formed in the prose).

mwachs5 reviewed (commented):
Is this adding semicolons to all the line endings? If so shouldn't all code examples be updated? Are there other changes you are suggesting here that should be updated in all the code examples?

Can you mark this appropriately as major/minor/patch?

mwachs5 commented:
In other words: are you trying to clarify the grammar or actually functionally change it? I wasn't sure from your PR description.

mmaloney-sf commented:
The changes here should parse the exact same `.fir` files as before. If it doesn't, that is an error.

seldridge reviewed (approved):
LGTM

This massively helps readability. It's also closer to the Python grammar specification. This isn't authoritative, but is another whitespace-sensitive language and is reasonable to model things off of.

mmaloney-sf commented on spec.md:
```diff
@@ -3873,27 +3873,51 @@ id = ( "_" | letter ) , { "_" | letter | digit_dec } | literal_id ;
+    "UInt" , [ width ]
+  | "SInt" , [ width ]
+  | "Analog" , [ width ] ;
+
```
I re-ordered everything as definition after use. This puts the gritty details of the grammar at the bottom and forefronts the high-level structure at the top. Eg, a FIRRTL design is a `circuit` and its made up of `decl`s (`module` `extmodule` `group` etc).

At the bottom, I picked a reasonable cut-off point and labeled all nonterminals below as "Token: ...". One small step towards a reasonable lex/parse distinction.

mmaloney-sf commented on spec.md:
```diff
@@ -3915,118 +3939,176 @@ primop_1expr2int_keyword =
+  | group
+  | skip ;
+
+(* Circuit Components *)
+circuit_component =
+    circuit_component_node
+  | circuit_component_wire
+  | circuit_component_reg
+  | circuit_component_inst
+  | circuit_component_mem ;
```
Yeah. I wanted to keep the one-line definitions together for readability, but I wanted the master list to be kept in order that they appear in the spec. I'm gonna keep it in the order here towards that end. Thank you though.

mmaloney-sf commented:
I did a bunch of whitespace cleanup and renamed the top-level declaration to `decl` (which means things like `module` are now `decl_module`... not 100% sure that's the right call, but going to see how it feels).

I also snuck in a definition for dynamic references `reference` is just the union of static and dynamic references, now renamed to `reference_static` and `reference_dynamic`. I also added a layer to separate `reference` from `expr_reference` in the `expr` tree. This is likely a material layer in the AST for any implementation.

mmaloney-sf commented on spec.md:
```diff
@@ -3873,27 +3873,51 @@ id = ( "_" | letter ) , { "_" | letter | digit_dec } | literal_id ;
+
+(* Enum Types *)
+type_enum = "{|" , { type_enum_alt } , "|}" ;
+type_enum_alt = id, [ ":" , type_constable ] ;
```
It's certainly reasonable in programming languages. It would amount to an empty type.

I'll put a note on my cleanup list to talk about this potentiality. Thanks for pointing this out.

dtzSiFive commented on spec.md:
```diff
@@ -3907,33 +3907,262 @@ said to have "internal" convention.
+
+(* References *)
+reference =
+    reference_static
```
I think this may be pre-existing, but I think this makes clear a bug?

Specifically, this doesn't appear to allow a subaccess in the middle of a chain of static referencing, specifically with trailing field (subfield) ?

mmaloney-sf replied:
Good catch.

This looks like it could be a bit tricky to resolve. I'll put this in my todos to tackle, and leave it for now.
